---
title: "Class 15: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Expression Analysis

We will use the DESeq2 package from bioconductor
```{r}
library(DESeq2)
```

Load our input coutData and colData files

```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

Hmmm... we need to remove that funky first **length** column from countData.

```{r}
countData <- countData[,-1]
head(countData)
```

```{r}
nonzero.inds <- rowSums(countData) != 0
countData <- countData[nonzero.inds, ]
head(countData)
```


## Running DESeq2

Nice now lets setup the DESeqDataSet object required for the DESeq() function and then run the DESeq pipeline. This is again similar to our last days hands-on session.

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)

```

```{r}
dds = DESeq(dds)
```

Get our results

```{r}
res <- results(dds)
head(res)
```

```{r}
summary(res)
```

# Volcono plot
Now we will make a volcano plot, a commonly produced visualization from this type of data that we introduced last day. Basically it's a plot of log2 fold change vs -log adjusted p-value.


```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="green", lwd=2)
abline(h=-log(0.01), col="green", lwd=2)
```


Add some color

```{r}
mycols <- rep("lightgray", nrow(res))
inds <- (abs(res$log2FoldChange) > 2 ) & res$padj < 0.01
mycols[ inds ] <- "blue"

plot(res$log2FoldChange, -log(res$padj), col=mycols)
abline(v=c(-2,2), col="black", lwd=2, lty=2)
abline(h=-log(0.01), col="black", lwd=2, lty=2)

```

```{r}
nrow(countData)
sum(inds, na.rm=TRUE)
```


## Adding gene annotation

Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

```

```{r}
res$symbol = mapIds(org.Hs.eg.db,        ## Annotation dataset
                    keys=row.names(res), ## Our existing IDs
                    keytype="ENSEMBL",   ## Format of our IDs
                    column="SYMBOL",     ## The format we want!
                    multiVals="first")   ## what to do for multiple matches
head(res)
```

Add some other gene IDs
```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```


## Section 2. Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

The **gage** package wants a vector as input with our numeric scores of interest with the vector names beting ENTREZ IDs

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```


```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```


```{r}
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)

Generate PDF foramts for these pathways

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

Now, let's process our results a bit more to automagicaly pull out the top 5 upregulated pathways, then further process that just to get the pathway IDs needed by the pathview() function. We'll use these KEGG pathway IDs for pathview plotting below.


```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

![](hsa04640.pathview.png)

![](hsa04630.pathview.png)

![](hsa00140.pathview.png)

![](hsa04142.pathview.png)

![](hsa04330.pathview.png)

## Gene ontology (GO)

Again we will use **gage()** but this time with GO gensets rather than KEGG

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

## Section 4. Reactome Analysis

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

