---
title: "Class 14: Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import our countData and colData into R

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```

Q1. How many genes are we dealing with here?

```{r}
nrow(counts)
```

Lets check if the cols of `counts` is thew same as the id column in `metadata`

```{r}
colnames(counts)
```

```{r}
metadata$id
```

```{r}
all( colnames(counts)[-1] == metadata$id )

```

Let's access just the control counts data so we can calculate an average counts value for each gene in these control samples.

```{r}
control <- metadata[metadata$dex=="control",]
control
```

Now we want to get the cols in counts with these `control$id` col-names

```{r}
control.mean <- rowSums( counts[ ,control$id] )/length(control$id) 
names(control.mean) <- counts$ensgene
head(control.mean)
```

Now do the same for the `treated` samples...

```{r}
treated <- metadata[metadata$dex=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/length(treated$id) 
names(treated.mean) <- counts$ensgene
```

Lets put these together into a new data.frame called `meancounts`

```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

```{r}
plot(meancounts)
```

Plot on log scale for both axis

```{r}
plot(meancounts, log="xy")
```

Calculate fold change and use log2 values

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. Again inspect your result (and the intermediate steps) to see if things make sense 

```{r}
#meancounts[,1:2]==0

# Identify the rows (i.e. genes) with zero values
to.rm <- unique(which( meancounts[,1:2]==0, arr.ind = TRUE )[,1])

# Remove these rows as we can't really say much about them.
mycounts <- meancounts[-to.rm,]
head(mycounts)

```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.
 
> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?
 
```{r}
up.ind <- mycounts$log2fc > 2
sum(up.ind)
```
 
```{r}
down.ind <- mycounts$log2fc < (-2)
sum(down.ind)
```

## 4. Adding annotation data

Read annotation data from ENSEMBLE...

```{r}
anno <- read.csv("annotables_grch38.csv")
```

We will use thge **merge()** function to add annotation data to our `mycounts` results

```{r}
mycounts.anno <- merge(x=mycounts, y=anno, by.x="row.names", by.y="ensgene")
head(mycounts.anno)
```

## We can use bioconductor to do this too...

```{r}
# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Hs.eg.db")
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(mycounts)
```


> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession as new columns called mycounts$entrez and mycounts$uniprot.

```{r}
mycounts$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

mycounts$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

head(mycounts)
```

# 5. DESeq2 analysis
Let's do this the right way. DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. 

```{r}
library("DESeq2")
```

```{r}
citation("DESeq2")
```

Build our deseq dataset object need fro running deseq analysis

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
#dds
```

```{r}
dds <- DESeq(dds)
```

Get our results

```{r}
res <- results(dds)
#as.data.frame(res)
```


```{r}
summary(res)
```

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```



```{r}
mycols <- rep("gray", nrow(res))
mycols[ res$padj < 0.05 ] <- "red"
mycols[ res$padj < 0.05 & abs(res$log2FoldChange) > 2 ] <- "blue"

plot( res$log2FoldChange,  -log(res$padj), col=mycols )
abline(h=log(0.05), col="darkgray", lty=2)
abline(v=c(-2,2), col="darkgray", lty=2)

```

Another way to plot this is with the **EnhancedVolcano**

But first lets put proper gene names on our results (i.e. gene 'symbols')

```{r}
x <- as.data.frame(res)
x$symbol <- mapIds(org.Hs.eg.db, 
                   keys=row.names(x),
                   keytype="ENSEMBL",
                   column="SYMBOL",
                   multiVals="first")


```

```{r}
library(EnhancedVolcano)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')  
```



