---
title: "Class 18: Cancer Genomics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Accessing the Genomic Data Commons

We need the bioconductor package **GenomicDataCommons**

```{r eval=FALSE}
BiocManager::install("GenomicDataCommons")
```

Let's use this package to ping the database.

```{r}
library(GenomicDataCommons)
status()
```

```{r}
library(TCGAbiolinks)

```


Query the GDC

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
```

```{r}
x <- cases_by_project$project.project_id
head(x)
```

```{r}
colvec <- rep("lightblue", nrow(x))

colvec[ x$key == "TCGA-PAAD" ] <- "red"

barplot(x$doc_count, names.arg = x$key, las=2, log="y", col=colvec)
```


# Section 2.

Step 1: Identify sequence regions that contain all 9-mer peptides that are only found in the tumor 
- What are the tumor specific amino-acids?
- What peptides include these amino acids?

```{r}
library(bio3d)

seqs <- read.fasta("lecture18_sequences.fa")
seqs
```

Does not look like these need to be aligned but we probably should align them to make sure when our input differs...

```{r}
aln <- seqaln(seqs)
aln
```

Lots of gap positions, lets remove them first

```{r}
aln.nogaps <- gaprm(aln)
aln.nogaps
```



```{r}
attributes(aln.nogaps)
```

```{r}
aln.nogaps$ali[,41]
```


```{r}
start.ind <- 41 - 8
end.ind <- 41 + 8

aln.nogaps$ali[, start.ind:end.ind ]
```


Find the sites (i.e. amino acids that are different)

```{r}
ide <- conserv(aln.nogaps, method = "identity")
mismatch.inds <- which(ide < 1)
```



```{r}
#seqs$ali[,tumor.sites]
ids <- paste( aln.nogaps$ali[1,mismatch.inds], 
       mismatch.inds,
       aln.nogaps$ali[2,mismatch.inds], sep="")
ids
```

```{r}
start.ind <-  mismatch.inds - 8
end.ind <-  mismatch.inds + 8

# Check for out-of-bounds numbers
min.pos <- 1
max.pos <- ncol(aln.nogaps$ali)

start.ind[start.ind < min.pos] <- min.pos
end.ind[end.ind > max.pos] <- max.pos


tumor <- NULL
for(i in 1:length(start.ind)) {
  tumor <- seqbind(tumor, 
                   aln.nogaps$ali[2,start.ind[i]: end.ind[i]])
}

write.fasta(ids=ids, ali=tumor, file="subsequences.fa")

```



gaprm() or gap.rm function to remove cols with gaps.
- should take vector, data.frame/matrix, alignment and pdbs object



```{r}


gaprm <- function(x) {
  #' Remove gap containing positions from alignment, pdbs 
  #'   or vector input objects
  #'
  #' @param x a character vector, fasta class alignment as obtained
  #'   from read.fasta() or a pdbs class object as obtained from
  #'   read.fasta.pdb()
  #'
  #' @return an object of similar class to the input but wwith gap
  #'   containing positions removed
  #'
  #' @examples gaprm( c("G","G","-","-","A") )

  is.fasta <- function(x) { inherits(x, "fasta") }

  gaps <- gap.inspect(x)

  message( paste("Dropping", length(gaps$t.inds),"gap positions and keeping", length(gaps$f.inds), "positions"))
  if(is.vector(x)) {
    x <- x[gaps$f.inds]
  }
  if(is.fasta(x)){
    x$ali <- x$ali[, gaps$f.inds, drop = FALSE]
  }
  if(is.pdbs(x)){
    x <- trim.pdbs(x, col.inds=gaps$f.inds)
  }
  return(x)
}
```


Or have the write.fasta() function have na.omit gap.omit option!

